File: src/webapp/src/model/marathon.ts
import { User } from './user';
import { Question } from './question';

export class Marathon {

  id: string;
  name: string;
  creator: User;
  startDate: Date;
  endDate: Date;
  submissionsStartDate: Date;
  submissionsEndDate: Date;
  description: string;
  onsite: boolean;
  location: string;
  language: string;
  maxGamesPerRunner: number;
  maxCategoriesPerGame: number;
  hasMultiplayer: boolean;
  maxNumberOfScreens: number;
  twitch: string;
  twitter: string;
  discord: string;
  youtube: string;
  country: string;
  discordPrivacy: boolean;
  submitsOpen: boolean;
  moderators: User[];
  defaultSetupTime: string;
  defaultSetupTimeHuman: string;
  selectionDone: boolean;
  scheduleDone: boolean;
  isPrivate: boolean;
  hasIncentives: boolean;
  canEditSubmissions: boolean;
  questions: Question[];
  hasDonations: boolean;
  payee: string;
  donationCurrency: string;
  supportedCharity: string;
  donationWebhook: string;
  donationsTotal: number;
  hasSubmitted: boolean;
  donationsOpen: boolean;
  videoRequired: boolean;
  unlimitedGames: boolean;
  unlimitedCategories: boolean;
  emulatorAuthorized: boolean;

  constructor() {
    this.onsite = false;
    this.maxGamesPerRunner = 5;
    this.maxCategoriesPerGame = 3;
    this.maxNumberOfScreens = 4;
    this.discordPrivacy = false;
    this.submitsOpen = false;
    this.selectionDone = false;
    this.donationsOpen = false;
    this.videoRequired = true;
    this.unlimitedGames = false;
    this.unlimitedCategories = false;
    this.emulatorAuthorized = true;
    this.isPrivate = true;
    this.questions = [];
  }
}

File: src/main/java/app/oengus/domain/marathon/Marathon.java
package app.oengus.domain.marathon;

import app.oengus.entity.model.Question;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.apache.commons.lang3.StringUtils;

import java.time.Duration;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@RequiredArgsConstructor
public class Marathon {
    private final String id;
    private final int creatorId;

    private String name;
    private ZonedDateTime startDate;
    private ZonedDateTime endDate;
    private ZonedDateTime submissionsStartDate;
    private ZonedDateTime submissionsEndDate;
    private String description;
    private boolean onsite = false;
    private String location;
    private String country;
    private String language = "en";
    private int maxGamesPerRunner = 5;
    private int maxCategoriesPerGame = 3;
    private boolean hasMultiplayer = true;
    private int maxNumberOfScreens = 4;
    private String twitch;
    private String twitter;
    private String mastodon;
    private String discord;
    private boolean discordPrivate = false;
    private boolean submissionsOpen = false;
    private Duration defaultSetupTime;
    private boolean selectionDone = false;
    private boolean scheduleDone = false;
    private boolean isPrivate = false;
    private boolean unlimitedGames = false;
    private boolean unlimitedCategories = false;
    private boolean emulatorAuthorized = true;
    private List<Integer> moderatorIds = new ArrayList<>();
    private boolean canEditSubmissions = false;
    // TODO: better model? Separate route?
    private List<Question> questions = new ArrayList<>();
    private String webhook;
    private String youtube;
    private boolean announceAcceptedSubmissions = false;

    // Submissions settings
    private boolean videoRequired = true;

    // Donation settings
    private boolean hasDonations = false;
    private boolean donationsOpen = true;
    private boolean hasIncentives = false;
    private String payee;
    private String supportedCharity;
    private String donationCurrency;

    // Discord settings
    private String discordGuildId;
    private String discordGuildName;
    private boolean discordRequired = false;

    // TODO: teams/volunteering settings

    public boolean hasWebhook() {
        return StringUtils.isNotEmpty(this.getWebhook());
    }
}
